compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirDataFlowAnalyzer.kt

```diff
@@ -899,8 +902,11 @@ abstract class FirDataFlowAnalyzer<FLOW : Flow>(
         }
 
         if (isAssignment) {
-            if (initializer is FirConstExpression<*> && initializer.kind == FirConstKind.Null) return
-            flow.addTypeStatement(propertyVariable typeEq initializer.typeRef.coneType)
+            if (initializer is FirConstExpression<*> && initializer.kind == FirConstKind.Null) {
+                flow.addTypeStatement(propertyVariable typeEq property.returnTypeRef.coneType.withNullability(ConeNullability.NULLABLE))
```

Comment:
> This change has no sense. In Flow we hold only types which somehow specifies type of some variable, so real type of variable is `intersection(originalType, typesFromFlow)`.

But, that's when type info in DFA is used to determine smartcast. From the point of view of `Flow`, this record, even though it could be just same as original type, is necessary.

> And now see how it can be read in some example:
> 
> ```
> var s: String? = "" // original type is String?
> s = null // we add statement that variable of type String? has type String?
> ```

Note that, the first line doesn't add anything since it's not `isAssignment`, but that's not the point here. If we change it a bit like:
```
var s: String?
s = "" // s -> [exactType: String]
s.length // smartcast
s = null // s -> [exactType: String?]
s.length // no smartcast
```

Needless to say, if a single persistent flow spans all expressions, we will see
* `s` is of `String`, smartcasts for both, if `s = null` is not recorded
* `s` is of `String?`, no smartcasts at all, if `s = null` is recorded

Splitting the flow will resolve the issue. Let's take a look at branches:

```
s = "" // flow@1: s -> [exactType: String]
when // or try
  ... -> // or catch
    s = null // flow@2: s -> [exactType: String?] with this change, otherwise, just same as above
}
// or finally, i.e. join point
// flow@3 = join(flow@1 , flow@2)
s.length // s is nullable!
```
For now, without this change, `s` is always `String`, since no additional info is recorded. Even with this change, due to the differential computation of join, if any incoming flow has no change for that variable, one type change in one branch is ignored together:
```
flow@3
= join(flow@1, flow@2)
= join(diff(flow@1, flow@1), diff(flow@1, flow@2))
= join(empty set, s -> ...)
= flow@1 again, so s = String, unsound smartcast
```

So, both of this `null` assignment record and element-wise join are needed to handle type info inside `finally` (or other similar join points). Expression _after_ `try` expression is a different story, though.